File: cpp/easy/reverse_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == NULL || head->next == NULL)
            return head;
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        ListNode* prev = new_head;
        ListNode* curr = head;
        ListNode* post = head->next;
        while(post != NULL)
        {
            curr->next = post->next;
            post->next = prev->next;
            prev->next = post;
            post = curr->next;
        }
        return new_head->next;
    }
};

Conclusion:
easy algorithm issue, time complexity is: O(n), space complexity is: O(1)

File: python/easy/reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        new_head = ListNode(0)
        new_head.next = head
        curr = head
        next = head.next
        while(next != None):
            tmp = next.next
            curr.next = tmp
            next.next = new_head.next
            new_head.next = next
            next = curr.next
        return new_head.next
        
